subsys/debug/cpu_load/cpu_load.c

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <debug/cpu_load.h>
+#include <shell/shell.h>
+#ifdef DPPI_PRESENT
+#include <nrfx_dppi.h>
+#else
+#include <nrfx_ppi.h>
+#endif
+#include <helpers/nrfx_gppi.h>
+#include <nrfx_timer.h>
+#include <hal/nrf_rtc.h>
+#include <hal/nrf_clock.h>
+#include <hal/nrf_power.h>
+#include <debug/ppi_trace.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(cpu_load, CONFIG_CPU_LOAD_LOG_LEVEL);
+
+/* Convert event address to associated publish register */
+#define PUBLISH_ADDR(evt) (volatile u32_t *)(evt + 0x80)
+
+/* Indicates that channel is not allocated. */
+#define CH_INVALID 0xFF
+
+/* Define to please compiler when periodic logging is disabled. */
+#ifdef CONFIG_CPU_LOAD_LOG_INTERVAL
+#define CPU_LOAD_LOG_INTERVAL CONFIG_CPU_LOAD_LOG_INTERVAL
+#else
+#define CPU_LOAD_LOG_INTERVAL 0
+#endif
+
+static nrfx_timer_t timer = NRFX_TIMER_INSTANCE(CONFIG_CPU_LOAD_TIMER_INSTANCE);
+static bool ready;
+static struct k_delayed_work cpu_load_log;
+static u32_t cycle_ref;
+static u32_t shared_ch_mask;
+
+#define IS_CH_SHARED(ch) \
+	(IS_ENABLED(CONFIG_CPU_LOAD_USE_SHARED_DPPI_CHANNELS) && \
+	(BIT(ch) & shared_ch_mask))
+
+
+/** @brief Allocate (D)PPI channel. */
+static nrfx_err_t ppi_alloc(u8_t *ch, u32_t evt)
+{
+	nrfx_err_t err;
+#ifdef DPPI_PRESENT
+	if (*PUBLISH_ADDR(evt) != 0) {
+		if (!IS_ENABLED(CONFIG_CPU_LOAD_USE_SHARED_DPPI_CHANNELS)) {
+			return NRFX_ERROR_BUSY;
+		}
+		/* Use mask of one of subscribe registers in the system,
+		 * assuming that all subscribe registers has the same mask for
+		 * channel id.
+		 */
+		*ch = *PUBLISH_ADDR(evt) & DPPIC_SUBSCRIBE_CHG_EN_CHIDX_Msk;
+		err = NRFX_SUCCESS;
+		shared_ch_mask |= BIT(*ch);
+	} else {
+		err = nrfx_dppi_channel_alloc(ch);
+	}
+#else
+	err = nrfx_ppi_channel_alloc((nrf_ppi_channel_t *)ch);
+#endif
+	return err;
+}
+
+static nrfx_err_t ppi_free(u8_t ch)
+{
+#ifdef DPPI_PRESENT
+	if (!IS_ENABLED(CONFIG_CPU_LOAD_USE_SHARED_DPPI_CHANNELS)
+		|| ((BIT(ch) & shared_ch_mask) == 0)) {
+		return nrfx_dppi_channel_free(ch);
+	} else {
+		return NRFX_SUCCESS;
+	}
+#else
+	return nrfx_ppi_channel_free((nrf_ppi_channel_t)ch);
+#endif
+}
+
+static void ppi_cleanup(u8_t ch_tick, u8_t ch_sleep)
+{
+	nrfx_err_t err = NRFX_SUCCESS;
+
+	if (IS_ENABLED(CONFIG_CPU_LOAD_ALIGNED_CLOCKS)) {
+		err = ppi_free(ch_tick);
+	}
+
+	if ((err == NRFX_SUCCESS) && (ch_sleep != CH_INVALID)) {
+		err = ppi_free(ch_sleep);
+	}
+
+	if (err != NRFX_SUCCESS) {
+		LOG_ERR("PPI channel freeing failed (err:%d)", err);
+	}
+}
+
+static void cpu_load_log_fn(struct k_work *item)
+{
+	u32_t load = cpu_load_get();
+	u32_t percent = load / 1000;
+	u32_t fraction = 1000 % percent;
```

Comment:
This is wrong. Should be `load % 1000`.
The same could be used in `cmd_cpu_load_get` as well.

Reply:
my bad.